### ISSUE TITLE
Possible to require semi colon only when needed?

### ISSUE THREAD

**Issue by trusktr** [palantir/tslint#2542]
For example,

This code is fine and doesn't need semi colons:

```ts
const p = props
const v = 'foo'
```

But this code may be error-prone:

```ts
const p = props
(window as any).test = "test"
```

because it is the same as 

```ts
const p = props(window as any).test = "test"
```

which tries to call `props` as a function.

Can we make semi colons only required when cases like that are detected, otherwise never require a semicolon? So we would write code like this:

```ts
const foo = 'foo' // no semicolon
const p = props; // semicolon required

(window as any).test = "test" // no semicolon
```

This would also teach people why semis are needed or not needed, rather than just tell them "always use semi colons".

Maybe I can achieve exactly this in a different way using other rules, like requiring function call parens on the same line as the function identifier? Any other cases that can be covered by rules, to achieve the same overall effect of requiring semis only in those special cases?

**Comment by buu700** [comment]
Looks like tslint-eslint-rules has [a rule that does exactly this](https://github.com/buzinas/tslint-eslint-rules/blob/master/src/docs/rules/noExtraSemiRule.md).

**Comment by ajafff** [comment]
@trusktr So you basically want something like eslint's [no-unexpected-multiline](http://eslint.org/docs/rules/no-unexpected-multiline)?

**Comment by trusktr** [comment]
Hey guys, yeah something like those would be great!

**Comment by buu700** [comment]
Apparently there's already `"semicolon": [true, "never"]` in tslint itself and no-unexpected-multiline is included in tslint-eslint-rules, so you can just combine those two and it'll do exactly what you want.

**Comment by trusktr** [comment]
Cool, it works!
